### ISSUE TITLE
setsockopt fails on joinAddr

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11108]
|[<img alt="ghazel's avatar" src="https://avatars.githubusercontent.com/u/31402?s=50" width="50" height="50">](https://github.com/ghazel)| @ghazel reported|
|-|-|
|Trac ID|trac#1578|
|Type|defect|
|Created|2006-03-16 02:23:05Z|

Granted that this traceback goes into multicast code in IOCP which has not been accepted yet, this clearly indicates that setsockopt could fail on any reactor on Windows, and the error case should be handled:

```
Traceback (most recent call last):
 File "C:\Python24\lib\site-packages\twisted\internet\defer.py", line 307, in _runCallbacks
   self.result = callback(self.result, *args, **kw)
 File "C:\Python24\lib\site-packages\twisted\internet\iocpreactor\udp.py", line 230, in _joinAddr1
   return self.reactor.resolve(interface).addCallback(self._joinAddr2, addr, join)
 File "C:\Python24\lib\site-packages\twisted\internet\defer.py", line 191, in addCallback
   callbackKeywords=kw)
 File "C:\Python24\lib\site-packages\twisted\internet\defer.py", line 182, in addCallbacks
   self._runCallbacks()
--- <exception caught here> ---
 File "C:\Python24\lib\site-packages\twisted\internet\defer.py", line 307, in _runCallbacks
   self.result = callback(self.result, *args, **kw)
 File "C:\Python24\lib\site-packages\twisted\internet\iocpreactor\udp.py", line 239, in _joinAddr2
   self.socket.setsockopt(socket.IPPROTO_IP, cmd, addr + interface)
 File "<string>", line 1, in setsockopt
socket.error: (10049, "Can't assign requested address")
```

<details><summary>Searchable metadata</summary>

```
trac-id__1578 1578
type__defect defect
reporter__ghazel ghazel
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1142475785000000 1142475785000000
changetime__1143337309000000 1143337309000000
version__ 
owner__ 
cc__itamarst
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst set owner to @itamarst|
|-|-|

One obvious solution is to replace socket.error with a meaningful twisted.internet.error.SomethingError. So, what was the problem that caused this? An invalid multicast address? lack of multicast support? Or is this a bug in the twisted code and joining that address ought to work?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @radix|
|-|-|

Ready for review in setsockopt-error-1578

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix set owner to @exarkun|
|-|-|

Reviewed +1 hooray yes errors

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `closed`|
|-|-|

(In [16454]) Merge setsockopt-error-1578

Authors: exarkun, itamarst
Reviewer: radix
Fixes [#11108](https://github.com/twisted/twisted/issues/11108)

This changes multicast support to catch errors from setsockopt and turn them 
into a new error, t.i.e.MulticastJoinError.

A test for this behavior is also added.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

(In [16455]) Merge iocp-multicast-1578

Author: ghazel
Reviewers: exarkun, itamarst
Fixes [#11108](https://github.com/twisted/twisted/issues/11108), [#11037](https://github.com/twisted/twisted/issues/11037)

Add multicast support to the IOCP reactor.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
